packages/mobile/src/escrow/saga.ts

```diff
@@ -102,6 +134,74 @@ function* transferStableTokenToEscrow(action: EscrowTransferPaymentAction) {
   }
 }
 
+function* transferStableTokenToEscrowWithoutCode(action: EscrowTransferPaymentAction) {
+  Logger.debug(TAG + '@transferToEscrowWithoutCode', 'Begin transfer to escrow')
+  try {
+    ValoraAnalytics.track(EscrowEvents.escrow_transfer_start)
+    const { phoneHashDetails, amount, context } = action
+    const { phoneHash, pepper } = phoneHashDetails
+    const [contractKit, walletAddress]: [
+      ContractKit,
+      string,
+      string,
+      FeelessVerificationState
+    ] = yield all([call(getContractKit), call(getConnectedUnlockedAccount)])
+
+    const [stableTokenWrapper, escrowWrapper]: [StableTokenWrapper, EscrowWrapper] = yield all([
+      call([contractKit.contracts, contractKit.contracts.getStableToken]),
+      call([contractKit.contracts, contractKit.contracts.getEscrow]),
+    ])
+
+    const escrowPaymentIds: string[] = yield call(
+      [escrowWrapper, escrowWrapper.getReceivedPaymentIds],
+      phoneHash
+    )
+
+    // Need to increment the derivation path of the paymentId based on
+    // how many are pending to avoid collisions with existing escrow txs
+    const addressIndex = escrowPaymentIds.length
+    const { publicKey, privateKey } = generateDeterministicInviteCode(pepper, addressIndex)
```

Comment:
Whenever a payment is redeemed, it is removed from the received array. As a result if a user receives an escrow payment with this methods, redeems it, and then received another the payment ID will be repeated. (In theory, the signature form the previous transaction can then be reused to avoid the need to know the recipients pepper.)
More immediately problematic though is that if the recipient had two pending payments, redeemed the first, then received a third, the second and third would have the same payment ID as the second and as a result would fail.

Reply:
Ok, so I have an idea for what we can do here.

When we are creating a payment, we can simply try derivation indices repeatedly, starting at 0 and incrementing, until we hit one that is not used. So the process here would be to:
1. Retrieve all the `escrowPaymentIds` for a recipient and turn them into a set.
2. Set `index = 0`
3. Generate `paymentId = generateDeterministicInviteCode(pepper, index)`
4. If `paymentId` exists in `escrowPaymentIds`, increment `index` and repeat step 3

Withdrawals can be accomplished with a similar process to find the private key for a given payment. In order to prevent looping forever in case there is a payment not generated by this process in the list, some cap such as `1000` should be used for the index.

Reply:
Astute observation. This wasn't on my radar because the way escrow currently works (redeem all at once and then never again) doesn't lead to this issue. However, there is no guarantee things will remain this way (in fact, we really should make the escrow experience better) so definitely better to get in front of this.

You definitely saved future me a lot of debugging time here!